onnx/test/cpp/inliner_test.cc

```diff
@@ -0,0 +1,164 @@
+  inliner::InlineLocalFunctions(model);
+  std::cout << ProtoToString(model) << "\n";
+  shape_inference::InferShapes(model);
+}
+
+TEST(FunctionInliner, BasicTest) {
+  const char* code = R"ONNX(
+<
+  ir_version: 8,
+  opset_import: [ "" : 10, "local" : 1 ]
+>
+agraph (float[N, 128] X, float[128,10] W, float[10] B) => (float[N, 10] C)
+{
+  T = local.foo (X, W, B)
+  C = local.square(T)
+}
+
+<
+  opset_import: [ "" : 10 ],
+  domain: "local",
+  doc_string: "Function foo."
+>
+foo (x, w, b) => (c) {
+  T = MatMul(x, w)
+  S = Add(T, b)
+  c = Softmax(S)
+}
+
+<
+  opset_import: [ "" : 10 ],
+  domain: "local",
+  doc_string: "Function square."
+>
+square (x) => (y) {
+  y = Mul (x, x)
+}
+)ONNX";
+
+  ModelProto model;
+  InlineFunctions(model, code);
+  auto num_nodes = model.graph().node_size();
+  ASSERT_EQ(num_nodes, 4);
+  auto num_functions = model.functions_size();
+  ASSERT_EQ(num_functions, 0);
+}
+
+TEST(FunctionInliner, Nested) {
+  const char* code = R"ONNX(
+<ir_version: 8, opset_import: [ "" : 17, "local" : 1 ]>
+agraph (float[N] X) => (float[N] Y)
+{
+  Y = local.foo (X)
+}
+
+<opset_import: [ "" : 17, "local" : 1 ], domain: "local">
+foo (x) => (y) {
+  temp = Add(x, x)
+  y = local.bar(temp)
+}
+
+<opset_import: [ "" : 17 ], domain: "local">
+bar (x) => (y) {
+  y = Mul (x, x)
+}
+)ONNX";
+
+  ModelProto model;
+  InlineFunctions(model, code);
+  auto num_nodes = model.graph().node_size();
+  ASSERT_EQ(num_nodes, 2);
+  auto num_functions = model.functions_size();
+  ASSERT_EQ(num_functions, 0);
+}
+
+TEST(FunctionInliner, Renaming) {
+  const char* code = R"ONNX(
+<ir_version: 8, opset_import: [ "" : 17, "local" : 1 ]>
+agraph (float[N] X) => (float[N] Y)
+{
+  temp = local.foo (X)
+  temp__1 = Mul (temp, temp)
+  Y = Abs (temp__1)
+}
+
+<opset_import: [ "" : 17, "local" : 1 ], domain: "local">
+foo (x) => (y) {
+  temp = Add(x, x)
+  y = Neg (temp)
+}
+)ONNX";
+
+  ModelProto model;
+  InlineFunctions(model, code);
+}
+
+TEST(FunctionInliner, OpsetMismatch) {
+  const char* code = R"ONNX(
+<ir_version: 8, opset_import: [ "" : 17, "local" : 1 ]>
+agraph (float[N] X) => (float[N] Y)
+{
+  temp = local.foo (X)
+  Y = local.bar (temp)
+}
+
+<opset_import: [ "" : 18], domain: "local">
+foo (x) => (y) {
+  y = Add(x, x)
+}
+
+<opset_import: [ "" : 17], domain: "local">
+bar (x) => (y) {
+  y = Add(x, x)
+}
+)ONNX";
+
+  ModelProto model;
+  InlineFunctions(model, code);
+
+  // The first node's call, to foo, must not be inlined.
+  auto& first_node = model.graph().node(0);
+  // Check that it is still a call to foo
+  ASSERT_EQ(first_node.op_type(), "foo");
+
+  // The second node's call, to bar, must be inlined.
+  auto& second_node = model.graph().node(1);
+  // Check that it is a call to Add
+  ASSERT_EQ(second_node.op_type(), "Add");
+
+  // The non-inlined foo must still be in the function list.
+  ASSERT_EQ(model.functions_size(), 1);
+  ASSERT_EQ(model.functions(0).name(), "foo");
+}
+
```

Comment:
I would add a test where the same function is added twice to check there is no duplicated name. And I would also add a test to check the inlining works well when a local function calls another local function (recursivity).

Reply:
There is already a test for one local function calling another local function.

I will add a test for same function being called twice.

Reply:
Done